tests/cases/conformance/types/union/discriminatedUnionTypes2.ts

```diff
@@ -0,0 +1,74 @@
+// @strict: true
+
+function f10(x : { kind: false, a: string } | { kind: true, b: string } | { kind: string, c: string }) {
+    if (x.kind === false) {
+        x.a;
+    }
+    else if (x.kind === true) {
+        x.b;
+    }
+    else {
+        x.c;
+    }
+}
+
+function f11(x : { kind: false, a: string } | { kind: true, b: string } | { kind: string, c: string }) {
+    switch (x.kind) {
+        case false:
+            x.a;
+            break;
+        case true:
+            x.b;
+            break;
+        default:
+            x.c;
+    }
+}
+
+function f13(x: { a: null; b: string } | { a: string, c: number }) {
+    x = { a: null, b: "foo", c: 4};  // Error
+}
+
+function f14<T>(x: { a: 0; b: string } | { a: T, c: number }) {
+    if (x.a === 0) {
+        x.b;  // Error
+    }
+}
+

```

Comment:
Could it be worth adding a test that includes constrained generics? My reasoning being that in the future it _might_ be worth considering discriminants that have constrained parameters because they are comparable to literals in some form. Having a test that would flag any changes might be helpful. Something like:

```ts
function f21<T extends number>(x: { a: undefined; b: { x: number } } | { a: T, b: { y: number } }) {
    if (x.a === undefined) {
        x.b = { y: 42 };
    }
}
```